In [10]:
import pandas as pd
import numpy as np

## Question 1

In [11]:
pd.__version__

'1.5.3'

In [12]:
!wget https://raw.githubusercontent.com/alexeygrigorev/datasets/master/housing.csv


--2023-09-18 14:49:37--  https://raw.githubusercontent.com/alexeygrigorev/datasets/master/housing.csv
Resolving raw.githubusercontent.com (raw.githubusercontent.com)... 185.199.108.133, 185.199.109.133, 185.199.110.133, ...
Connecting to raw.githubusercontent.com (raw.githubusercontent.com)|185.199.108.133|:443... connected.
HTTP request sent, awaiting response... 200 OK
Length: 1423529 (1.4M) [text/plain]
Saving to: ‘housing.csv.3’

housing.csv.3       100%[===================>]   1.36M  --.-KB/s    in 0.01s   

2023-09-18 14:49:37 (111 MB/s) - ‘housing.csv.3’ saved [1423529/1423529]



In [13]:
df = pd.read_csv("/content/housing.csv")
df.head(5)

longitude  latitude  housing_median_age  total_rooms  total_bedrooms  \
0    -122.23     37.88                41.0        880.0           129.0   
1    -122.22     37.86                21.0       7099.0          1106.0   
2    -122.24     37.85                52.0       1467.0           190.0   
3    -122.25     37.85                52.0       1274.0           235.0   
4    -122.25     37.85                52.0       1627.0           280.0   

   population  households  median_income  median_house_value ocean_proximity  
0       322.0       126.0         8.3252            452600.0        NEAR BAY  
1      2401.0      1138.0         8.3014            358500.0        NEAR BAY  
2       496.0       177.0         7.2574            352100.0        NEAR BAY  
3       558.0       219.0         5.6431            341300.0        NEAR BAY  
4       565.0       259.0         3.8462            342200.0        NEAR BAY

## Question 2

In [14]:
print("Number of columns: {}".format(len(df.columns)))

Number of columns: 10


## Question 3

In [15]:
# Verificar qué columnas tienen valores faltantes
columnas_con_faltantes = df.columns[df.isna().any()].tolist()

# Mostrar las columnas con valores faltantes
print("Columnas con valores faltantes:", columnas_con_faltantes)

Columnas con valores faltantes: ['total_bedrooms']


In [16]:
df.isnull().sum()

longitude               0
latitude                0
housing_median_age      0
total_rooms             0
total_bedrooms        207
population              0
households              0
median_income           0
median_house_value      0
ocean_proximity         0
dtype: int64

## Question 4

In [17]:
df.nunique()

longitude               844
latitude                862
housing_median_age       52
total_rooms            5926
total_bedrooms         1923
population             3888
households             1815
median_income         12928
median_house_value     3842
ocean_proximity           5
dtype: int64

### ocean_proximity has 5 unique values

## Question 5

In [18]:
df.groupby("ocean_proximity").median_house_value.mean()

ocean_proximity
<1H OCEAN     240084.285464
INLAND        124805.392001
ISLAND        380440.000000
NEAR BAY      259212.311790
NEAR OCEAN    249433.977427
Name: median_house_value, dtype: float64

### The average value of the median_house_value for the houses located near the bay is : 259212.3

## Question 6

In [19]:
media = df["total_bedrooms"].mean()
print("Average of total_bedrooms before changes")
print(media)

Average of total_bedrooms before changes
537.8705525375618


In [20]:
# Rellenar los valores faltantes en la columna 'C' usando la estrategia de relleno 'ffill' (propagar hacia adelante)
df['total_bedrooms'].fillna(media, inplace=True)

In [21]:
print("Average of total_bedrooms after changes")
print(df["total_bedrooms"].mean())

Average of total_bedrooms after changes
537.8705525375617


## Question 7

In [22]:
filtro1 = (df["ocean_proximity"]=="ISLAND")
df[filtro1]

longitude  latitude  housing_median_age  total_rooms  total_bedrooms  \
8314    -118.32     33.35                27.0       1675.0           521.0   
8315    -118.33     33.34                52.0       2359.0           591.0   
8316    -118.32     33.33                52.0       2127.0           512.0   
8317    -118.32     33.34                52.0        996.0           264.0   
8318    -118.48     33.43                29.0        716.0           214.0   

      population  households  median_income  median_house_value  \
8314       744.0       331.0         2.1579            450000.0   
8315      1100.0       431.0         2.8333            414700.0   
8316       733.0       288.0         3.3906            300000.0   
8317       341.0       160.0         2.7361            450000.0   
8318       422.0       173.0         2.6042            287500.0   

     ocean_proximity  
8314          ISLAND  
8315          ISLAND  
8316          ISLAND  
8317          ISLAND  
8318          ISLAND

In [23]:
df2 = df[["housing_median_age", "total_rooms", "total_bedrooms"]]
df2

housing_median_age  total_rooms  total_bedrooms
0                    41.0        880.0           129.0
1                    21.0       7099.0          1106.0
2                    52.0       1467.0           190.0
3                    52.0       1274.0           235.0
4                    52.0       1627.0           280.0
...                   ...          ...             ...
20635                25.0       1665.0           374.0
20636                18.0        697.0           150.0
20637                17.0       2254.0           485.0
20638                18.0       1860.0           409.0
20639                16.0       2785.0           616.0

[20640 rows x 3 columns]

In [24]:
data = df2[filtro1]
data

housing_median_age  total_rooms  total_bedrooms
8314                27.0       1675.0           521.0
8315                52.0       2359.0           591.0
8316                52.0       2127.0           512.0
8317                52.0        996.0           264.0
8318                29.0        716.0           214.0

In [25]:

X  = np.array(data)
X

array([[  27., 1675.,  521.],
       [  52., 2359.,  591.],
       [  52., 2127.,  512.],
       [  52.,  996.,  264.],
       [  29.,  716.,  214.]])

In [26]:
def vector_vector_multiplication(u, v):
    assert u.shape[0] == v.shape[0]

    n = u.shape[0]

    result = 0.0

    for i in range(n):
        result = result + u[i] * v[i]

    return result

def matrix_vector_multiplication(U, v):
    assert U.shape[1] == v.shape[0]

    num_rows = U.shape[0]

    result = np.zeros(num_rows)

    for i in range(num_rows):
        result[i] = vector_vector_multiplication(U[i], v)

    return result

def matrix_matrix_multiplication(U, V):
    assert U.shape[1] == V.shape[0]

    num_rows = U.shape[0]
    num_cols = V.shape[1]

    result = np.zeros((num_rows, num_cols))

    for i in range(num_cols):
        vi = V[:, i]
        Uvi = matrix_vector_multiplication(U, vi)
        result[:, i] = Uvi

    return result

In [27]:
transpuesta = X.T
transpuesta

array([[  27.,   52.,   52.,   52.,   29.],
       [1675., 2359., 2127.,  996.,  716.],
       [ 521.,  591.,  512.,  264.,  214.]])

In [28]:
XTX = matrix_matrix_multiplication(transpuesta, X)
XTX

array([[9.6820000e+03, 3.5105300e+05, 9.1357000e+04],
       [3.5105300e+05, 1.4399307e+07, 3.7720360e+06],
       [9.1357000e+04, 3.7720360e+06, 9.9835800e+05]])

In [29]:

inv = np.linalg.inv(XTX)
inv

array([[ 9.19403586e-04, -3.66412216e-05,  5.43072261e-05],
       [-3.66412216e-05,  8.23303633e-06, -2.77534485e-05],
       [ 5.43072261e-05, -2.77534485e-05,  1.00891325e-04]])

In [30]:
y = np.array([950, 1300, 800, 1000, 1300])
y

array([ 950, 1300,  800, 1000, 1300])

In [31]:
z = matrix_matrix_multiplication(inv, transpuesta)
z

array([[-0.00825608, -0.00653208, -0.00232159,  0.02565144,  0.01204934],
       [-0.00165852,  0.0011141 ,  0.00139656, -0.00103215, -0.00110698],
       [ 0.00754365, -0.00301964, -0.00455125,  0.00181685,  0.00329418]])

In [32]:
w = matrix_vector_multiplication(z, y)
w

array([23.12330961, -1.48124183,  5.69922946])

### The last value is 5.69

---

